### PR TITLE
Add keyword reporting for publishers

### PR DESCRIPTION
**Pull request by ericholscher**
This builds a new KeywordImpression index,
which is used to show ads matching a specific keyword.
This is a somewhat naive approach,
but I think it's useful to know how any ad matches a keyword,
and I can't think of a way to not show 100%+ for those cases.

### CODE REVIEW

File: adserver/views.py [hunk 1]
```diff
@@ -562,11 +562,6 @@ def get_context_data(self, **kwargs):
         start_date = self.get_start_date()
         end_date = self.get_end_date()
         campaign_type = self.request.GET.get("campaign_type", "")
-        revenue_share_percentage = self.request.GET.get("revenue_share_percentage", "")
-        div_id = self.request.GET.get("div_id", "")
-        # This needs to be something other than `advertiser` to not conflict with template context on advertising reports.
-        report_advertiser = self.request.GET.get("report_advertiser", "")
-        country = self.request.GET.get("country", "")
```
Comment (ericholscher) [inline adserver/views.py:569]: These were moved off the base onto the specific reports.

File: adserver/tasks.py [hunk 2]
```diff
@@ -143,3 +148,48 @@ def daily_update_impressions(day=None):
...
+
+    for impression_type in IMPRESSION_TYPES:
+        queryset = _default_filters(impression_type, start_date, end_date)
+
+        for values in (
+            queryset.values("publisher", "advertisement", "keywords")
+            .annotate(total=Count("keywords"))
+            .filter(total__gt=0)
+            .order_by("-keywords")
+        ):
```
Comment (davidfischer) [inline adserver/tasks.py:172]: This gets the unique publisher/advertisement/keywords combination for a day. This is going to be enormously expensive especially for publishers like Read the Docs. Did you try this in production to see how long it takes and how many results it yields? I am worried about running this every 30 minutes.

File: adserver/tasks.py [hunk 3]
```diff
@@ -143,3 +148,48 @@ def daily_update_impressions(day=None):
...
+
+            keywords = json.loads(values["keywords"])
+            ad = Advertisement.objects.filter(id=values["advertisement"]).first()
```
Comment (davidfischer) [inline adserver/tasks.py:175]: Since you use the flight later, you could save a query (in the loop) with a `select_related` here.

**Review by davidfischer — APPROVED** [review]
This looks fine. My big worry is the performance due to all the combinatorics.

File: adserver/tasks.py (hunk 2 above)
Comment (ericholscher) [reply adserver/tasks.py:172]: It runs hourly, but we could also scale that back if we're worried about it. I haven't tested it against RTD.

Comment (ericholscher) [reply adserver/tasks.py:172]: I tested it with today's date in prod and it ran in ~10 seconds.

File: adserver/tasks.py (hunk 3 above)
Comment (ericholscher) [reply adserver/tasks.py:175]: Thinking more on it, I think we might be able to return this data from the query. I'll see if that works reasonably well.

File: adserver/tasks.py (hunk 2 above)
Comment (ericholscher) [reply adserver/tasks.py:172]: Removing the inline queries makes it really fast 👍

**Comment by davidfischer** [comment]
Your improvements here look good. Removing the queries in the loop is great.

**Pull request merged**